Add action keywords for each plugin

gordonwatts opened the pull request:
## Summary of the Pull Request

Added keywords for each plug-in the PowerToys Run module.

The keywords proposed are:

- Calculator: =
- Indexer: ?
- Program: .
- URI: //
- Window Walker: <

All the plug-ins that were a member of the global plug-in list (used "*" as an action keyword in the json configuration) retain that.

## PR Checklist
* [x] Applies to #5823
* [x] CLA signed
* [x] Tests added/passed
* [x] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where a discussion took place: #5823. I was not directly involved in the discussion, other than reading it. This might impact a PR that is trying to remove the space between the action keyword and the search string (see 

## Info on Pull Request

See discussion in issue #5823. Changes most `plugin.json` files in the PowerToys Runner plug in to use `actionKeywords` rather than `actionKeyword`, and adds a non-global keyword as noted above.

### Other Changes:

Two things were needed to make action keywords work properly. Hopefully, they do not impact other operations of PowerToys!

- The WindowsWalker plugin was using `RawQuery` instead of `Query` to drive its searches. `RawQuery`, of course, contains the action keyword (like ":"). Fixed by changing `RawQuery` to `Query`.
- `QueryBuilder` adds all global plug-ins to the dictionary of plug-in list for each query. When a plugin is a member of both global and nonglobal lists, it can be added twice to the plugin-pair dictionary. This behavior is protected with a simple if statement to prevent double insertion.

### Oustanding Issues:

Need advice: the current settings infrastructure stores the keywords in the user's settings file. This will override what changes are done here. Normally this is fine - however, there is no UI for altering the keywords. Someone who has already run `PowerToys` and wants to take advantage of this will have to edit the settings json file by hand.

## Validation Steps Performed

Ran the tests and also tested by hand

- Some tests failed before (and after) I made modifications: the fancyzones tests and the rename tests.

gordonwatts committed:
Add action keywords for each plugin

- Added keywords for each plug-in
-  Change window walker search to use Query, not RawQuery (to get rid of action keyword)
- Make sure plug-ins that have action keywords as well as are in global list don't get added twice

All tests passed that passed before I started

Addressing issue #5823

gordonwatts commented:
Failed tests seem to be the rename - which looks like it is failing across many many tests not just mine. I am not sure how to mark as "ready for review" - but this does need input/help at the very least from maintainers.

htcfreek commented:
> I am not sure how to mark as "ready for review"

A PR is ready to review if it is not marked as draft.

gordonwatts commented:
> @gordonwatts
> 
> > ## PR Checklist
> > 
> > * [x]  Applies to [f64df97](https://github.com/microsoft/PowerToys/commit/f64df976528c78e5f75a8fc04fffa588f5375afc)
> 
> Note: Normally the corresponding issue and not an commit id is named here.
> 
> > I am not sure how to mark as "ready for review"
> 
> A PR is ready to review if it is not marked as draft.
> You can define reviewers. And if you are not allowed to do so, you can write a comment and name the reviewers.

Thanks!

- I fixed the `Applies To`
- I don't know anyone on the project, but looking at issue #5823, I would suggest @crutkas for review (I can't define reviewers myself in this repo).

crutkas commented:
rerunning CI, failure was transient

crutkas commented:
CI passed.  We're getting toward end of our 0.21 release window.  Lets target this for 0.22 release window as i want to have more time to try this

gordonwatts commented:
Sounds great! Let me know what you need from me.

crutkas commented:
Playing around with this, liking it but the indexer didn't want to work.  On 0.22 it does work and on master.

General rules, if a file can use it, i don't think it should a hot key, example & since & will pull up snip & sketch other apps.

https://stackoverflow.com/questions/1976007/what-characters-are-forbidden-in-windows-and-linux-directory-names

**Like:**
- Calculator: =
   - Exact lead you'd expect here

**Unsure of hotkey:**
- Indexer: /
   - `?` as you're querying the system.

- URI: &
   - `:` as http:// ...  It mimics a keyshortcut that you'd already be used to.

- Running Process (Window Walker): :
   - `<` since the shell plugin uses >.  Feels a bit ok.

- Program: $
   - Thoughts is i want this to be quick, this is one of our main use cases
   - `#` feels more programmy
   - `@` could maybe feel right here too
   - '/' is super quick at least on a US keyboard.

**Don't think we need:**
- Folder: \
   - why: c:\ already pulls, unless i'm missing something, i couldn't get this to pop up with

crutkas commented:
@gordonwatts looking at your PR, really don't see why indexer wouldn't work

gordonwatts commented:
I noticed this last night when I tried the indexer for the first time. I thought I'd tested it. I'll get all the commits from master and then have a go at it. Actually, I see there are now conflicts, so I need to do this anyway, I suppose. I'll take your advice on keyboard shortcuts as well when I do that.

gordonwatts committed:
Merge remote-tracking branch 'upstream/master' into issue5823

gordonwatts committed:
Fix bad formatting from the merge

gordonwatts committed:
Do not add global plug-ins if we have an action keyword

Note side effect: you will not not be able to search for things that start with one of the plug-in keywords.

gordonwatts committed:
Missed merge conflict repair!

gordonwatts committed:
Revert "Do not add global plug-ins if we have an action keyword"

This reverts commit e3b0ecd708455c28b4ebf1a590ab5bdb1990f21a.

gordonwatts commented:
This is an update:

- Pulled all changes from master, and resolved conflicts.
- The folder and indexer plug-in suddenly started working!

Need to investigate (mostly note to myself):

- [ ] Why `$ github` gets the "run command: github" as top billing rather than the "Github Desktop" program

Your proposed keyboard shortcuts are fine. I'll update those when I've figured out what is going on above.

Apologies - I have an early morning code review shift for my work so more tomorrow!

crutkas commented:
@gordonwatts, bet you got some weird point in time where maybe something broke it and we fixed it in a later commit

gordonwatts commented:
## Why `$ github` always has the run command rising to the top

This took me a while (slightly embarrassed). Also, it doesn't always have it rise to the top, it turns out. Sometimes it is second in the list.

- The program plugin actually does two things:
   - Looks for matches in UWP and win32 install programs
   - Looks for exact matches to run a program (like the .exe name).

In the case that there is an installed program with the same name as a .exe, they both get the same score. And then they do not sort in a stable way.

Example: If I type "$ github" I get back three results from the Program plugin: "run command: github", "Application: github", "Application: Github Desktop". The first two both get a score of 172, and the third gets 164 (I'm pretty sure those numbers are accurate).

I think the unstable sort here is a bug, but it isn't a bug for this PR. We can open another issue and fix it there if you think that is something that should be fixed. Now that I understand this behavior - I'm going to move onto getting the keys updated to your suggestions.

gordonwatts committed:
Adjust keywords to suggested ones

gordonwatts commented:
I have adjusted the keywords in the last push, and I checked that they were all working. I think those are the updates you asked me for, @crutkas.

One other thing - in order for the system to see the new settings in the plugin.json files, I had to delete my current settings file in `C:\Users\xxx\AppData\Local\Microsoft\PowerToys\PowerToys Run\Settings\Settings.json`. However, I notice that when the new version re-creates this file, the action keywords are no longer in there (nice!). I'm not sure what got updated, and I don't know how this will affect users. But if they have anything in their settings files w.r.t. keywords, then they will not see these changes.

gordonwatts committed:
No need to ubild a query if we aren't going to use it

Didn't quite get the if statement right the first time I put it in.

damnskippy commented:
Hello, I had this suggestion elsewhere in https://github.com/microsoft/PowerToys/issues/6691 to have different dedicated key activation shortcuts for launching powertoys-runs in different modes for searching.
- windows-only
- windows-of-currently-active-app (to switch among windows/instances of same app)
- files-only
- apps-only (shell plugin)
- open-url
- all-of-the-above (current behavior)

Maybe this is what is already proposed above, it's not clear to me, so adding this suggestion here FWIW.
Being able to search among windows of current app would be very useful to have, please consider.

gordonwatts commented:
I like some of these! The "currently active app" might be especially cool, if it could do tabs in chrome/edge (I see that the tabs will appear in the atl-tab display at some point for edge).

I'd suggest those become separate requests, however, rather than adding to this PR (depending on the feelings of the project coordinators, of course).

The one thing that might impact this is if it is better to have a different windows-level keyboard shortcut instead of the keys you enter here. In that case, this should be closed, and we should start from scratch.

damnskippy commented:
It would be significantly more efficient and quicker to be able to directly go into a specific search mode, than a 2 step process to first activate the app and then to go into a desired mode and then do the search.  It may seem there's just one extra key stroke, but it's more than that practically speaking.

damnskippy commented:
In pure window switching mode, it would be even better to list the user-chosen n number of most recently used windows in MRU order, as opposed to the current approach where the search window pops up blank or with the last search string.  All of this makes for a fast, efficient text based alt-tab switcher on steroids.  Just another thought.

gordonwatts commented:
> In pure window switching mode, it would be even better to list the user-chosen n number of most recently used windows in MRU order, as opposed to the current approach where the search window pops up blank or with the last search string. All of this makes for a fast, efficient text based alt-tab switcher on steroids. Just another thought.

I'd been thinking about this as another feature suggestion - a MRU that was filtered by whatever active searches there are currently. But I was going to make that a separate feature, rather than in this one. And I really wanted more experience with whatever this one turns out before I could figure out what I wanted next to improve this. So we need @crutkas to suggest the right way forward here.

crutkas commented:
@gordonwatts / @damnskippy , i think that would be a neat feature but no one called it out in spec review or i don't even think we have a tracking ask for that.  We're trying to be very focused on building what is being asked for and what the heart of the product.

I think for plugin management as well, being able to directly pull up different modes via keystrokes would be helpful as well.

crutkas commented:
I just grabbed it and seems to be quirky.  `?` seems to work reliably but the others don't.

![image](https://user-images.githubusercontent.com/1462282/94625673-2d9f3a00-026e-11eb-891b-30d1ffd3f27d.png)
![image](https://user-images.githubusercontent.com/1462282/94625726-4e678f80-026e-11eb-84d6-0cbbc2e5f280.png)

htcfreek commented:
I think in general we should support combinations of two or more characters as action keyword too. - With only using one character per plugin we don't have much possibilities for different action keywords.

And we need a general (template) solution if we want to have a setting which can be changed by the user.

gordonwatts commented:
> I just grabbed it and seems to be quirky. `?` seems to work reliably but the others don't.
> 
> ![image](https://user-images.githubusercontent.com/1462282/94625673-2d9f3a00-026e-11eb-891b-30d1ffd3f27d.png)
> ![image](https://user-images.githubusercontent.com/1462282/94625726-4e678f80-026e-11eb-84d6-0cbbc2e5f280.png)

For the first case, the ":o" - that is expected with the current behavior. If you add the " " between the ":" and the "o" then it will work as I think you are expecting it to.

I left this in place because there was a huge discussion in the PR for removing the space - what if these characters were meant to be part of a search string (to sum it up).

Personally, I'd be happy to disable that so that if you used an action key in front then it would disable the global search plug-in's, space or no.

For the second... "< outlook", I see this if outlook is running:
![image](https://user-images.githubusercontent.com/1778366/94637286-2b0c0700-02a5-11eb-8991-2ef09e64425b.png)

if I then kill outlook, and just type "alt-space", I get the following:

![image](https://user-images.githubusercontent.com/1778366/94637346-50991080-02a5-11eb-8785-9cc63c0d5d00.png)

That is, I think, a MRU bug, but then as soon as I type "< outlook", I get:

![image](https://user-images.githubusercontent.com/1778366/94637375-673f6780-02a5-11eb-91a7-a3256c9b099f.png)

which is expected.

In your case, when you type "< outlook" I assume that you have outlook running. A test of where the bug is (since I can't repro it here), is if you just type "outlook" - does the outlook window show up as expected?

gordonwatts commented:
> @gordonwatts, @crutkas and @damnskippy:
> 
> I think in general we should support combinations of two or more characters as action keyword too. - With only using one character per plugin we don't have much possibilities for different action keywords.
> 
> And we need a general (template) solution if we want to have a setting which can be changed by the user.

Currently, the JSON infrastructure allows you to have a keyword trigger an arbitrary number of plug-ins. So you could make "?" trigger both "local machine search/indexer" and also "web indexer" by just setting up the configuration correctly.

The current parser does not detect multiple action keywords in a row. As a feature request once this goes in that sounds like it might get some up-votes! I think what I'd do, if I had lots and lots of time (hahahahaha!) was first get this PR in, then get the UI in (see #5273). Once that was done, take a step back and see if there was still a need for multi-key presses.

crutkas commented:
1. My assumption would be if we trigger a action keyword, that it only searches that plugins reg'ed with that action word.
   1. if i type `<outlo` or `< outlo`, i only see results from Window Walker's plugin.
   1. typing `:o` shouldn't have returned anything since that was for the URI plugin.  why did search return stuff?
2. for the space for it work, i thought @alekhyareddy28 added in logic to not need the space for this but maybe that was localized only to the program plugin.

gordonwatts commented:
The space logic is pretty complicated, actually. Are there tests surrounding this?

The ":o" returns items from lots of different searches on my machine.

I will add logic so that if you have a search action key, only those search plug-ins will work. So ":o" and ": o", and "<outlook" and "< outlook" will always return the same thing.

BTW, did you sort out why "<outlook" didn't return anything? Or a clue as to what might be happening? That is a bug I can't repo on my local machine.

alekhyareddy28 commented:
@crutkas, the space requirement after the action keyword has been removed for all the plugins and it's not specific to the program plugin. Tests have been added to the query builder class.

Also, I'm just a bit concerned with adding action keywords for each plugin because without the action keyword the plugin won't work. So if we plan to add '=' as an action keyword for the calculator plugin 5+1 would not invoke the calculator plugin anymore. Anyone who is used to the current way of working of PT Run may not expect that.

crutkas commented:
@alekhyareddy28 isn't that what the * means?  i thought that would query all the time

for calculator, if i do 
"ActionKeywords": [ "*", "=" ], 

![image](https://user-images.githubusercontent.com/1462282/94868798-bc3fc280-03f8-11eb-8c0c-75cdb55c1c79.png)
![image](https://user-images.githubusercontent.com/1462282/94868811-c235a380-03f8-11eb-8d3f-b15930a31967.png)

alekhyareddy28 commented:
@crutkas, sorry I might have misunderstood the intent of this PR. I just looked at the description and not at the files modified.

When we inherited the code base from wox, plugins are of two types, global plugins (which have the action keyword as '*') and non-global plugins (Which have any action keyword other than '*'). We might want to validate that everything works as expected with this change because right now with the infrastructure that we have, global and non-global plugins are mutually exclusive. I'm not sure if these changes might cause some unexpected behaviour. I can take a deeper look at it over the weekend. 

 Also we definitely have to make code changes to prevent duplicate queries from being sent in the Query builder class.

gordonwatts commented:
> @crutkas, sorry I might have misunderstood the intent of this PR. I just looked at the description and not at the files modified. I thought the action keyword was modified to '=' from '*'.
> 
> When we inherited the code base from wox, plugins are of two types, global plugins (which have the action keyword as * ) and non-global plugins (Which have any action keyword other than '*'). We might want to validate that everything works as expected with this change because right now with the infrastructure that we have, global and non-global plugins are mutually exclusive. I'm not sure if these changes might cause some unexpected behaviour. I can take a deeper look at it over the weekend.
> 
> Also we definitely have to make code changes to prevent duplicate queries from being sent in the Query builder class.

Thanks for the comments @alekhyareddy28!

You are correct, without modification, the same query plug-in will be added to the list twice. There is an if statement in this PR to protect from that.

Further, from @crutkas 's comments, it sounds like it might be preferable to turn off all global plug-ins when a actionkey is detected. I agree with him, that makes the UI more sensable.

I was aiming to do that last night, but... exhaustion set in before I was able to tackle it.

I did not modify the query builder, so I did not modify QueryBuilder other than that one modification. I missed that there were tests there. That is my mistake. I will fix that.

So my plan:

- Modify the QB so that when an action keyword is detected, only that plug-in is run and no global plug-ins are added to the list.
- Add tests so check that functionality works.

I'm, obviously, open to suggestions. I've been using these features for a while now in my own build, and they have really changed the way I work with the launcher. I pretty much never use the global features any longer. So I'm feeling fairly motivated to get this or something like this in.

gordonwatts committed:
Do not add global plugins if we have an action keyword

This means, especially without a space, that global plugins do not become part of the search. Which really narrows down the search results, nicely.

gordonwatts commented:
Finally got back to this - apologies (helping to run a workshop this week).

- I can't figure out how to add tests to `QueryBuilderTest`. The crux of the problem is that QB references the global `PluginManager.GlobalPlugins`, and I don't see a way to use the `Moq` framework to shim something in. I'm not an expert at that framework, so I might be missing something. But I'll need help there (I looked on SO and in the PluginManagerTest file). Further, I would note that none of the logic that scans and inserts items in the query plugin list for a query in QB is tested - only the logic that makes the input string uniform. Let me know how to proceed here.
- No longer add global plugins to list if an action keyword is matched.

That last change makes this a lot nicer, btw. I'm glad we could do that!

crutkas commented:
/azp run

crutkas commented:
/azp run

crutkas commented:
Playing with this, feels a lot more solid.

I have two requests that i feel make this natural.

- for URL-> do `//`  I tried this and if felt more natural and it works.
- for programs, i didn't like `/` since it was too folder / URL like. (yes i know i suggested it)
   - Tried `.` and `:`
   - Honestly, i was liking the period for quick app search.

Thoughts?  A lot is, what feels intutative and natural.  I sync'ed with Craig from WSL and he felt they all seemed natural.

@ryanbodrug-microsoft  / @alekhyareddy28  / @enricogior  would love your opinion.

htcfreek commented:
@crutkas 
What do you think about using the ampersand `&` character for program/app search? - Like the call operator in Powershell.

crutkas commented:
i find & to be a stretch keystroke.  happy to get more opinions.

gordonwatts commented:
I like the `//` idea. I'll wait to update for the conclusion of the search for app. I've been using it for a while, and I have to say, my fingers "learned" it. From my POV, I think once we choose something, our fingers will memorize it.

alekhyareddy28 commented:
@crutkas, I agree with the Program plugin. I think `/` does not feel very intuitive to use for the program plugin. I think the others are fine and we could get used to them once we start using those action keywords.

Regarding the PR, since we're not executing the global plugins once we see an action keyword, I wanted to point out that we should be mindful of adding action keywords which might be a part of a valid query in the future. We might not get a few results  because of this. 

@gordonwatts I'll test out this PR and review it sometime today. Thank you for this!
Also, regarding testing the QueryBuilder you could pass the global plugins as an argument, like how the non-global plugins are being passed and that could be mocked/set to what we want while testing.

gordonwatts commented:
Ok. Just to make sure we are on the same page here. I should change:

- The URL plug-in to "//"
- The program plug-in to "."

For the testing, @alekhyareddy28, the thing that needs to be tested is how it handles the global plug-ins after dealing with the non-global plug-ins. That logic doesn't really come into play unless the global plug-in list is built, and that isn't something that is passed in as an argument - the routine pulls it from a global variable. I looked a little bit into the moq framework to see if it could do global variable injection, but it didn't look like it. However, my knowledge of this framework is pretty minimal, so it is quite possible I missed something.

crutkas commented:
@gordonwatts yes, please change that.

alekhyareddy28 reviewed (commented):
I did try out the changes in this PR and they look good to me. 

A few comments - 
1. @crutkas, `.` seems to be a valid regex character that is valid for the indexer and ww plugin. On adding `.` as an action keyword we would not see those results anymore. I'm not sure if anyone is extensively using them but just wanted to point that out.
![image](https://user-images.githubusercontent.com/28739210/96187894-9f27eb00-0ef2-11eb-834c-47447ffbcc89.png)

So previously .txt would show up all the txt files but they won't show up once we set the action keyword of the program plugin to `.`

2. @gordonwatts It would be great to have a few tests to ensure that we don't break any functionality in the future as well. 
We could analyze the plugin-query pairs that are returned by the `Build` function to see how the global plugins are handled. You could convert the `Build` function so as to take the `GlobalPlugins` as an argument. This should fix the issue and it should be easy to add tests for the same. This would make it quite similar to how the non-global plugins are being handled right now.

Modify it to something like -
```
public static Dictionary<PluginPair, Query> Build(ref string text, Dictionary<string, PluginPair> nonGlobalPlugins, List<PluginPair> GlobalPlugins)
```
This way you can pass the GlobalPlugins and then analyze the plugin-query pairs to see how the global plugins were handled.

Once the tests are in and the action keywords are updated to // and . as suggested by crutkas the PR should be good to get in. Thank you!

alekhyareddy28 commented:
**Clarification regarding the above comment**. I was wrong about the period being treated as a regex by the indexer plugin. Indexer does not treat the Period as a regex. Even ww seems to have it's own scoring mechanism which just compares characters.
But presently as stated in the above comment, the user can search for all txt files using `.txt`. However, if we set the action keyword of the program plugin to `.`, when the user enters `.txt`, it is now treated as a search for `txt` within only the program plugin and may not return any results depending on the programs. The indexer results will not show up anymore. The rest of the plugin action keywords should not be an issue as they are restricted characters within the indexer plugin.

```
// Reserved keywords in oleDB
private readonly string reservedStringPattern = @"^[\/\\\$\%]+$|^.*[<>].*$";
```

gordonwatts commented:
Thanks, @alekhyareddy28, for the careful look!

1. I made the changes to the shortcuts (it is easy enough to change them back), and updated the header of this PR.
1. Indeed, ".txt" does not get you a text file. "?.txt" gets you the right thing.
1. Given we can have 2 character keywords, we could use that feature to work around this. I'm open to suggestions. One nice thing about working on this project is I do not have to make a final decision here. It is good not to be in charge once in a while. ;-)
1. The tests are a good idea - and what you describe @alekhyareddy28 sounds perfect. I'd not realized that modifying the API was an option. I'll do that. However, I can't get to that before the end of the weekend or early next week. I'm planning on taking a bit of a much-needed break this weekend. Apologies for not being able to get to it sooner.

gordonwatts committed:
Update keywords as requested in the PR

alekhyareddy28 commented:
Issue #7378 captures the remaining work related to unit testing this functionality.

enricogior commented on src/modules/launcher/Plugins/Microsoft.Plugin.Folder/plugin.json:
```diff
@@ -1,6 +1,6 @@
-    "ActionKeyword":"*",
+    "ActionKeyword": "*",
     "Name":"Folder",
     "Description":"Searcn and open folders",
```
Typo: Searcn -> Search

damnskippy commented:
@crutkas @gordonwatts Any considerations/thoughts on being able invoke specific modes with a direct keyboard shortcut as suggested/discussed above?  Just curious.  Thanks.

crutkas commented:
@damnskippy, something discussed in https://github.com/microsoft/PowerToys/issues/5273
